### PR TITLE
Add integration test for node registration methods

### PR DESCRIPTION
**Pull request by andrejtokarcik**
An integration test for the node registration logic introduced in #5182. Changes needed to allow for logger injection/capture are also included inasmuch as this is the first test that analyzes logs emitted by the Teleport process.

### CODE REVIEW

File: integration/integration_test.go [hunk 1]
```diff
@@ -4854,6 +4830,86 @@ func (s *IntSuite) TestBPFSessionDifferentiation(c *check.C) {
...
+			return tconf
+		}
+		t.StartNodeAndRegisterOnPort(nodeConfig(), tc.getRegisterPort(t))
+		c.Assert(t.StopAll(), check.IsNil)
+
+		capturedLog := s.w.LogBuf.String()
+		c.Logf("Test Case %d - %s", i, tc.comment)
+		c.Log(capturedLog)
+		matched, _ := regexp.MatchString(tc.logPattern, capturedLog)
+		c.Assert(matched, check.Equals, true,
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply integration/integration_test.go:4908]: Didn't notice this, thanks!

File: lib/auth/register.go [hunk 2]
```diff
@@ -114,6 +115,14 @@ type HostCredentials func(context.Context, string, bool, RegisterUsingTokenReque
 // tokens to prove a valid auth server was used to issue the joining request
 // as well as a method for the node to validate the auth server.
 func Register(params RegisterParams) (*Identity, error) {
+	return RegisterWithLogger(params, log)
+}
+
+// RegisterWithLogger allows to specify a custom logger and capture logs
+// written during the registration process, primarily for testing purposes.
+func RegisterWithLogger(params RegisterParams, logger logrus.FieldLogger) (*Identity, error) {
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply lib/auth/register.go:123]: My original reason was that `logger` didn't pertain to the registration logic parameters as such. But now that you mention it I see `RegisterParams` already contains similar fields (e.g. `Clock`) so I'll move the logger there as well.

File: integration/integration_test.go [hunk 3]
```diff
@@ -4854,6 +4830,86 @@ func (s *IntSuite) TestBPFSessionDifferentiation(c *check.C) {
...
+			tconf.SSH.Enabled = true
+
+			return tconf
+		}
+		t.StartNodeAndRegisterOnPort(nodeConfig(), tc.getRegisterPort(t))
+		c.Assert(t.StopAll(), check.IsNil)
+
+		capturedLog := s.w.LogBuf.String()
+		c.Logf("Test Case %d - %s", i, tc.comment)
+		c.Log(capturedLog)
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply integration/integration_test.go:4906]: Yep, my idea was to not rely on the `tearDownTest` cleanups to print the failure on error but to manage them explicitly to enable the log content checks. Note that `c.Log`  [adds the data only to the test error output](https://pkg.go.dev/gopkg.in/check.v1#C.Log).

File: integration/integration_test.go [hunk 4]
```diff
@@ -4854,6 +4830,86 @@ func (s *IntSuite) TestBPFSessionDifferentiation(c *check.C) {
...
+			"registration via proxy is preferred over auth when proxy default port is specified",
+			func(i *TeleInstance) string {
+				return fmt.Sprintf("%v", defaults.HTTPListenPort)
+			},
+			`(?s)Registering node to the cluster(.*)Attempting registration via proxy server(.*)Attempting registration with auth server`,
+		},
+	}
+
+	for i, tc := range testCases {
+		s.setUpTest(c)
```
[the remark this answers is not in the document]
Comment (andrejtokarcik) [reply integration/integration_test.go:4875]: Yes, I realize my approach is idiosyncratic. I considered your alternative with multiple functions but the table-driven format seemed to be more readable & in line with the conventions in place (i.e., to group related test cases within a single function).

I'll distribute it over multiple functions then, it should indeed require less ad hoc sophistry.

**Comment by andrejtokarcik** [comment]
Thanks for your thorough review @a-palchikov!

File: integration/integration_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (awly) [reply integration/integration_test.go:4875]: Please use https://github.com/gravitational/teleport/blob/master/rfd/0001-testing-guidelines.md for all new tests
we're trying to migrate away from gocheck.

File: integration/integration_test.go [hunk 5]
```diff
@@ -4854,6 +4830,86 @@ func (s *IntSuite) TestBPFSessionDifferentiation(c *check.C) {
...
+	var testCases = []struct {
+		comment         string
+		getRegisterPort func(i *TeleInstance) string
+		logPattern      string
+	}{
+		{
+			"successful registration with auth server",
+			func(i *TeleInstance) string {
+				return i.GetPortAuth()
+			},
```
Comment (awly) [inline integration/integration_test.go:4848]: you can set `getRegisterPort: i.GetPortAuth,` directly

File: integration/integration_test.go [hunk 6]
```diff
@@ -4854,6 +4830,86 @@ func (s *IntSuite) TestBPFSessionDifferentiation(c *check.C) {
...
+// TestNodeRegistrationMethods tests registration of a node
+// with auth server and via proxy.
+func (s *IntSuite) TestNodeRegistrationMethods(c *check.C) {
+	tr := utils.NewTracer(utils.ThisFunction()).Start()
+	defer tr.Stop()
+
+	var testCases = []struct {
+		comment         string
+		getRegisterPort func(i *TeleInstance) string
+		logPattern      string
```
Comment (awly) [inline integration/integration_test.go:4842]: matching against logs is fragile
instead, fetch the list of all nodes from the API (something like `t.GetSiteAPI().GetNodes()`) and check `node.GetUseTunnel()` to see if it connected over a tunnel (proxy) or auth.

Comment (andrejtokarcik) [reply integration/integration_test.go:4842]: The motivation here is to cover the changes introduced in #5182, namely to assure that:
(a) users are not presented with unnecessary `ERROR` level messages as in #4663,
(b) the proxy mode is attempted as the first registration method in certain scenarios (which does not preclude falling back to auth) while "auth first, fall back to proxy" remains the default.

Merely verifying the end state would not achieve these goals. Checking the logs seemed to be the most straightforward way to do that.

Comment (awly) [reply integration/integration_test.go:4842]: can you intercept connection attempts on the server or the client in any way?
like a wrapper around `net.Dialer` or `net.Listener`?

we're creating a bit of maintenance burden by checking logs with regexps, especially considering how slow and flaky our integration tests are.
next time someone changes the log message format, they will likely see integration tests fail (after ~15min), re-run them 1-2 more times (another ~30min) before looking at the errors and actually updating the test.

**Pull request closed without merging**